<a href="https://colab.research.google.com/github/JosephWildey/MachineLearningModels/blob/main/Multiple_Linear_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Multiple Linear Regression

## import libraries

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## import the dataset

In [ ]:
dataset = pd.read_csv("/home/joe/Documents/ML-Resources/50_Startups.csv")
x = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

## Encoding Categorical Data

Keep in mind the country data will switch from the final column to the first column

In [ ]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [-1])], remainder='passthrough')
x = np.array(ct.fit_transform(x))

# Splitting the dataset into the Training dataset and Test dataset

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(x, y, test_size = 0.2, random_state = 1)

## Training the Multiple Linear Regression Model on the Training set

In [ ]:
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, Y_train)

LinearRegression()

## Predicting the Test Set Results

Import caveat: This dataset has several features, not just one. Up to now everything has followed the same methodology as having one feature.

Instead of attempting to graph a 5-D plot, we will display two vectors. The first vector will be the real profit in the test set. There will also be a vector to show predicted results.

The step to get the first vector is the same as to get the prediction in Simple Linear Regression.

The final step will be to display the first vector and predicted vector results side-by-side using Numpy. Numpy has a lot of methods to handle arrays making it a natural choice for this goal.The method is concatenate which allows the concatenation of either vertically or horizontally two vectors or even arrays.

The step of combining and displaying two vectors starts with print, so we can print the output to the screen. Within print we can call the concatenate method of numpy.The concatenate method requires two vectors of the same shape (eg. 3x1, 3x1 or 1x3, 1x3). Also, they must be in parentheses because the concatenate method expects a tuple of vectors to concatenate.

Reshape is an attribute function of numpy concatenate that allows you to reshape vectors vertically or horizontally. We use len because the function requires the amount of columns in y_pred. The same idea can then be applied to the second vector, Y_test.

Concatenate can take two arguments: 0 or 1. 0 means you want to a vertical concatenation, and 1 means you want to do a horizontal concatenation.

In [ ]:
y_pred = regressor.predict(X_test)
np.set_printoptions(precision=2)
print(np.concatenate((y_pred.reshape(len(y_pred), 1), Y_test.reshape(len(Y_test), 1)), 1))

[[114664.42 105008.31]
 [ 90593.16  96479.51]
 [ 75692.84  78239.91]
 [ 70221.89  81229.06]
 [179790.26 191050.39]
 [171576.92 182901.99]
 [ 49753.59  35673.41]
 [102276.66 101004.64]
 [ 58649.38  49490.75]
 [ 98272.03  97483.56]]


## Making a Single Prediction

In [ ]:
print(regressor.predict([[1, 0, 0, 160000, 130000, 300000]]))

[180892.25]


This means according to our model that the profit of a Californian startup which spent 160,000 in R&D, 130,000 in Admin, and $300,000 in marketing was $180,892.25.

We use a double array because the predict function always expects a 2d array. 

Dummy variables are the first three values, not the last three.

## Getting the final linear regression equation with the values of the coefficients

In [ ]:
print(regressor.coef_)
print(regressor.intercept_)

[-2.85e+02  2.98e+02 -1.24e+01  7.74e-01 -9.44e-03  2.89e-02]
49834.885073230675


Profit = -285 x Dummy State 1 + 298 x Dummy State 2 - 12.4 x Dummy State 3 + 0.774 x R&D Spend - 0.00944 x Admin 0.0289 x Marketing + 49834.89